### PR TITLE
Add new decoration for SPV_INTEL_function_pointers

### PR DESCRIPTION
**Pull request by KornevNikita**
Spec: [pull/4958](https://github.com/intel/llvm/pull/4958)

### CODE REVIEW

**Comment by KornevNikita** [comment]
@AlexeySotkin @MrSidims @AlexeySachkov @mlychkov take a look please

File: lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -1282,6 +1282,41 @@ void SPIRVToLLVM::addMemAliasMetadata(Instruction *I, SPIRVId AliasListId,
...
+  std::vector<SPIRVDecorate const *> ArgumentAttributes =
+      BV->getDecorations(internal::DecorationArgumentAttributeINTEL);
+
+  for (const auto *Dec : ArgumentAttributes) {
+    std::vector<SPIRVWord> Literals = Dec->getVecLiteral();
+    SPIRVWord ArgNo = Literals[0];
+    SPIRVWord Attr = Literals[1];
+
+    AttrBuilder Builder;
+    if (Attr == FunctionParameterAttributeByVal)
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.cpp:1295]: If only one attr is expected per decoration then it would be better to use `switch`.

Comment (AlexeySotkin) [reply lib/SPIRV/SPIRVReader.cpp:1295]: That's right, one attribute per decoration.

File: test/transcoding/SPV_INTEL_function_pointers/decor-func-ptr-arg-attr.ll [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+!opencl.used.optional.core.features = !{!4}
+!opencl.compiler.options = !{!4}
+!llvm.ident = !{!5}
+
+!0 = !{i32 1, !"wchar_size", i32 4}
+!1 = !{i32 7, !"frame-pointer", i32 2}
+!2 = !{i32 1, i32 2}
+!3 = !{i32 4, i32 100000}
+!4 = !{}
+!5 = !{!"Cpmpiler"}
```
Comment (mlychkov) [inline test/transcoding/SPV_INTEL_function_pointers/decor-func-ptr-arg-attr.ll:65]:
```suggestion
!5 = !{!"Compiler"}
```

**Review by mlychkov — APPROVED** [review]
Other than a minor comment LGTM. Thank you.

File: lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -1282,6 +1282,49 @@ void SPIRVToLLVM::addMemAliasMetadata(Instruction *I, SPIRVId AliasListId,
...
+void transFunctionPointerCallArgumentAttributes(SPIRVValue *BV, CallInst *CI) {
+  std::vector<SPIRVDecorate const *> ArgumentAttributes =
+      BV->getDecorations(internal::DecorationArgumentAttributeINTEL);
+
+  for (const auto *Dec : ArgumentAttributes) {
+    std::vector<SPIRVWord> Literals = Dec->getVecLiteral();
+    SPIRVWord ArgNo = Literals[0];
+    SPIRVWord Attr = Literals[1];
+
+    switch (Attr) {
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.cpp:1294]: Can we use `SPIRSPIRVFuncParamAttrMap` instead of the `switch`?

File: lib/SPIRV/SPIRVWriter.cpp [hunk 4]
```diff
@@ -1984,6 +1984,32 @@ bool LLVMToSPIRVBase::shouldTryToAddMemAliasingDecoration(Instruction *Inst) {
...
+void addFuncPointerCallArgumentAttributes(CallInst *CI,
+                                          SPIRVValue *FuncPtrCall) {
+  auto AddDecorArgAttr{[&](int ArgNo, spv::FunctionParameterAttribute Attr) {
+    FuncPtrCall->addDecorate(
+        new SPIRVDecorate(spv::internal::DecorationArgumentAttributeINTEL,
+                          FuncPtrCall, ArgNo, Attr));
+  }};
+
+  for (unsigned ArgNo = 0; ArgNo < CI->arg_size(); ++ArgNo) {
+    if (CI->paramHasAttr(ArgNo, Attribute::ByVal))
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVWriter.cpp:1996]: Can we use `SPIRSPIRVFuncParamAttrMap` instead of multiple `if`s?

File: lib/SPIRV/SPIRVReader.cpp [hunk 5]
```diff
@@ -1282,6 +1282,26 @@ void SPIRVToLLVM::addMemAliasMetadata(Instruction *I, SPIRVId AliasListId,
...
 }
 
+void transFunctionPointerCallArgumentAttributes(SPIRVValue *BV, CallInst *CI) {
+  std::vector<SPIRVDecorate const *> ArgumentAttributes =
+      BV->getDecorations(internal::DecorationArgumentAttributeINTEL);
+
+  for (const auto *Dec : ArgumentAttributes) {
+    std::vector<SPIRVWord> Literals = Dec->getVecLiteral();
+    SPIRVWord ArgNo = Literals[0];
+    SPIRVWord SpirAttr = Literals[1];
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.cpp:1292]:
```suggestion
    SPIRVWord SpirvAttr = Literals[1];
```

**Review by AlexeySotkin — APPROVED** [review]

**Pull request merged**